### PR TITLE
Add more hir APIs for associated items

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by matklad** [comment]
bors d=SomeoneToIgnore

File: crates/ra_hir/src/code_model.rs [hunk 1]
```diff
@@ -639,17 +644,49 @@ pub struct MacroDef {
...
+    fn as_assoc_item(self, db: &impl DefDatabase) -> Option<AssocItem> {
+        as_assoc_item(db, AssocItem::TypeAlias, self.id)
+    }
+}
+fn as_assoc_item<ID, DEF, CTOR, AST>(db: &impl DefDatabase, ctor: CTOR, id: ID) -> Option<AssocItem>
+where
+    ID: Lookup<Data = AssocItemLoc<AST>>,
+    DEF: From<ID>,
+    CTOR: FnOnce(DEF) -> AssocItem,
+    AST: AstNode,
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]: I have when `where` is bigger than the function body TT

**Comment by SomeoneToIgnore** [comment]
Almost LGTM, I need the same `container` method returning `AssocItemContainer` implemented for `Function` due to https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_hir/src/code_model.rs#L1031 closure returning a `Function`

**Comment by matklad** [comment]
function.as_assoc_item().unwrap().container() should work I think. It's a
shame that we need to unwrap here, but I think explaining in the type
system that this function is also an associated item is not worth the
trouble.

**Comment by SomeoneToIgnore** [comment]
Works indeed, just thought that it's simple to add such method to `Function` also.
By the way, I was able to import that trait automatically 🎉

File: crates/ra_hir/src/code_model.rs [hunk 2]
```diff
@@ -769,6 +818,7 @@ impl TypeParam {
     }
 }
 
+// FIXME: rename to `ImplBlock`
```
Comment (kjeremy) [inline crates/ra_hir/src/code_model.rs]: It already is?

Comment (matklad) [reply crates/ra_hir/src/code_model.rs]: s/to/from, it should be Impl :D

Comment (SomeoneToIgnore) [reply crates/ra_hir/src/code_model.rs]: I'll adjust this in my other PR, seems not to be worth of a separate PR.

**Pull request merged**